src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile_isolated_strategy.py

```diff
@@ -5,33 +5,201 @@
 from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
                         unicode_literals, with_statement)
 
-import itertools
 import os
-import shutil
-import uuid
-from collections import OrderedDict, defaultdict
+import Queue as queue
+from collections import OrderedDict, defaultdict, namedtuple
 
 from pants.backend.jvm.tasks.jvm_compile.jvm_compile_strategy import JvmCompileStrategy
 from pants.backend.jvm.tasks.jvm_compile.resource_mapping import ResourceMapping
 from pants.base.build_environment import get_buildroot
-from pants.base.worker_pool import Work
+from pants.base.exceptions import TaskError
+from pants.base.worker_pool import Work, WorkerPool
 from pants.util.dirutil import safe_mkdir, safe_walk
 
 
+UNSTARTED = 'Unstarted'
+SUCCESS = 'Success'
+FAILURE = 'Failure'
+QUEUED = 'Queued'
+
+class ExecutionGraph(object):
+
+  class ExecutionFailure(Exception):
+    """Raised when tasks fail during execution"""
+
+  class Work(namedtuple('ExecutionWork', ['fn', 'dependencies', 'on_success', 'on_failure'])):
+    def __call__(self, *args, **kwargs):
+      self.fn()
+
+    def run_success_callback(self):
+      if self.on_success:
+        self.on_success()
+
+    def run_failure_callback(self):
+      if self.on_failure:
+        self.on_failure()
+
+  class StatusTable(object):
```

Comment:
Ok. My sense for when to use an inner class in python isn't quite there yet.
wrt moving it to it's own file
That's the plan. I wasn't sure if I should do that with this change, or wait until I introduce using this into other places.